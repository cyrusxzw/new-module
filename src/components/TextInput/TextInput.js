import React from 'react';
import PropTypes from 'prop-types';
import styles from './TextInput.module.css';
import cx from 'classnames';

const TextInput = ({
  className,
  defaultValue,
  hasContent,
  hasError,
  id,
  isEnabled,
  inputClassName,
  inputRef,
  label,
  max,
  min,
  name,
  type,
  maxLength,
  hasValidation,
  onBlur,
  onChange,
  prefixElement,
  theme,
  value,
  ...moreProps
}) => {
  const dataTestRef = moreProps['data-test-ref'];
  const classNames = cx(styles.formText, className);
  const inputClassNames = cx(
    styles.formTextInput,
    styles[theme],
    (!!defaultValue || hasContent) && styles.hasContent,
    !!hasError && styles.hasError,
    inputClassName,
  );
  const labelClassSet = cx(styles.formTextLabel, styles[theme]);

  return (
    <div className={classNames}>
      <label htmlFor={id}>
        <input
          className={inputClassNames}
          data-test-ref={dataTestRef}
          defaultValue={defaultValue}
          disabled={!isEnabled}
          id={id}
          max={max}
          maxLength={maxLength}
          min={min}
          name={name}
          noValidate={!hasValidation}
          onBlur={onBlur}
          onChange={onChange}
          ref={inputRef}
          type={type}
          value={value}
        />
        <span className={labelClassSet}>{label}</span>
        {prefixElement}
      </label>
    </div>
  );
};

TextInput.propTypes = {
  className: PropTypes.string,
  defaultValue: PropTypes.string,
  hasContent: PropTypes.bool,
  hasError: PropTypes.bool,
  hasValidation: PropTypes.bool,
  id: PropTypes.string.isRequired,
  inputClassName: PropTypes.string,
  inputRef: PropTypes.func,
  isEnabled: PropTypes.bool,
  label: PropTypes.string,
  max: PropTypes.number,
  min: PropTypes.number,
  name: PropTypes.string,
  maxLength: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  prefixElement: PropTypes.element,
  theme: PropTypes.oneOf(['dark', 'light']),
  type: PropTypes.oneOf([
    'text',
    'password',
    'email',
    'number',
    'tel',
    'search',
  ]),
  'data-test-ref': PropTypes.string,
  value: PropTypes.string,
};

TextInput.defaultProps = {
  className: undefined,
  defaultValue: undefined,
  hasContent: false,
  hasError: false,
  hasValidation: false,
  inputClassName: undefined,
  inputRef: undefined,
  isEnabled: true,
  label: undefined,
  max: undefined,
  min: undefined,
  name: undefined,
  maxLength: undefined,
  onBlur: undefined,
  onChange: undefined,
  prefixElement: undefined,
  theme: 'dark',
  type: 'text',
  'data-test-ref': undefined,
  value: undefined,
};

export { TextInput };
