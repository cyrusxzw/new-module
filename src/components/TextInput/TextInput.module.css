@import 'styles/index';

.formText {
  width: 100%;
}

.formTextInput {
  position: relative;
  width: 100%;
  box-sizing: border-box;
  padding-left: 10px;
  border: 1px solid var(--color-white);
  background-color: transparent;
  border-radius: 3px;
  caret-color: var(--color-white);
  color: var(--color-white);
  font-family: var(--font-family-sans-serif-regular);
  font-size: rem(20px);
  line-height: 1.85;

  &.dark {
    border-color: var(--color-dark-copy);
    caret-color: var(--color-dark-copy);
    color: var(--color-dark-copy);
  }

  &.light {
    border-color: var(--color-light-copy);
    caret-color: var(--color-light-copy);
    color: var(--color-light-copy);
  }

  &.hasError {
    border-color: var(--color-error);
  }

  &:focus {
    outline: none;
  }

  /* Overrides for autofill coloring */
  &:-webkit-autofill,
  &:-webkit-autofill:hover,
  &:-webkit-autofill:focus {
    &.dark {
      -webkit-text-fill-color: var(--color-dark-copy);
    }
    &.light {
      -webkit-text-fill-color: var(--color-light-copy);
    }

    transition: background-color 5000s ease-in-out 0s;
  }

  @media (--viewport-md-only) {
    font-size: rem(17px);
    line-height: 2.18;
  }
  @media (--viewport-xs-sm-only) {
    font-size: rem(18px);
    line-height: 2.068;
  }
}

.formTextLabel {
  position: absolute;
  top: 2px;
  left: 10px;
  font-family: var(--font-family-sans-serif-regular);
  font-size: rem(20px);
  line-height: 1.85;
  opacity: 1;
  pointer-events: none;
  transition: all 300ms var(--easing-ease-in-out-cubic);

  &.dark {
    color: var(--color-dark-copy);
  }

  &.light {
    color: var(--color-light-copy);
  }

  &.is-hidden {
    opacity: 0;
  }

  .formTextInput:focus + &,
  .formTextInput.hasContent + & {
    opacity: 0;
  }
  @media (--viewport-md-only) {
    font-size: rem(17px);
    line-height: 2.18;
  }
  @media (--viewport-xs-sm-only) {
    font-size: rem(18px);
    line-height: 2.068;
  }
}

/*
.formTextErrorMessage {
  color: var(--color-error);
} */
