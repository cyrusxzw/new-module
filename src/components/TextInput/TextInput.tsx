import React from 'react';
import styles from './TextInput.module.css';
import cx from 'classnames';
import type { TextInputType } from './TextInput.type';

const TextInput: TextInputType = ({
  className,
  defaultValue,
  hasContent = false,
  hasError = false,
  id,
  inputClassName,
  inputRef,
  isEnabled = true,
  label,
  max,
  maxLength,
  min,
  name,
  onBlur,
  onChange,
  prefixElement,
  theme = 'dark',
  type = 'text',
  value,
  ...moreProps
}) => {
  const dataTestRef = moreProps['data-test-ref'];
  const classNames = cx(styles.formText, className);
  const inputClassNames = cx(
    styles.formTextInput,
    styles[theme],
    (!!defaultValue || hasContent) && styles.hasContent,
    !!hasError && styles.hasError,
    inputClassName,
  );
  const labelClassSet = cx(styles.formTextLabel, styles[theme]);

  return (
    <div className={classNames}>
      <label htmlFor={id}>
        <input
          className={inputClassNames}
          data-test-ref={dataTestRef}
          defaultValue={defaultValue}
          disabled={!isEnabled}
          id={id}
          max={max}
          maxLength={maxLength}
          min={min}
          name={name}
          onBlur={onBlur}
          onChange={onChange}
          ref={inputRef}
          type={type}
          value={value}
        />
        <span className={labelClassSet}>{label}</span>
        {prefixElement}
      </label>
    </div>
  );
};

export { TextInput };
