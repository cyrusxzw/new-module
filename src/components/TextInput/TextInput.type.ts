import type { ChangeEvent, FocusEvent, Ref, ReactElement } from 'react';
import type { ComponentWithoutChildren, Themes } from '~/types';

type TextInputProps = {
  'data-test-ref'?: string;
  className?: string;
  defaultValue?: string;
  hasContent?: boolean;
  hasError?: boolean;
  /** @deprecated no underlying functionality */
  hasValidation?: boolean;
  id: string;
  inputClassName?: string;
  inputRef?: Ref<HTMLInputElement>;
  isEnabled?: boolean;
  label?: string;
  max?: number;
  maxLength?: number;
  min?: number;
  name?: string;
  onBlur?: (event: FocusEvent<HTMLInputElement>) => void;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
  prefixElement?: ReactElement;
  theme?: Themes;
  type?: 'text' | 'password' | 'email' | 'number' | 'tel' | 'search';
  value?: string;
};

type TextInputType = ComponentWithoutChildren<TextInputProps>;

export { TextInputProps, TextInputType };
