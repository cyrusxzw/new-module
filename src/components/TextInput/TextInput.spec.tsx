import React from 'react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';
import { TextInput } from './TextInput';

describe('<TextInput />', () => {
  it('should render base component correctly', () => {
    const { container } = render(
      <TextInput data-test-ref="test-data-ref" id="test-textinput" />,
    );

    expect(container).toMatchSnapshot();
  });

  it('should be accessible with a correct label', async () => {
    const { container } = render(
      <TextInput
        data-test-ref="test-data-ref"
        id="test-textinput"
        label="label"
      />,
    );

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it.skip('should be accessible with a placeholder attribute', async () => {
    const { container } = render(
      <TextInput data-test-ref="test-data-ref" id="test-textinput" />,
    );

    /** @TODO This component needs to be refactored to correctly (not)render label html */
    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });
});
