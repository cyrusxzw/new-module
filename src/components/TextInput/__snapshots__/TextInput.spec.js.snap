// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> should render base component correctly 1`] = `
<div>
  <div
    class="formText"
  >
    <label
      for="test-textinput"
    >
      <input
        class="formTextInput dark"
        data-test-ref="test-data-ref"
        id="test-textinput"
        novalidate=""
        type="text"
        value=""
      />
      <span
        class="formTextLabel dark"
      />
    </label>
  </div>
</div>
`;
